images/virtualization-artifact/pkg/controller/vi/internal/source/object_ref_vd.go

```diff
@@ -0,0 +1,361 @@
+		condition.Message = ""
+
+		var imageStatus virtv2.ImageStatusSize
+		imageStatus.Stored = vdRef.Spec.PersistentVolumeClaim.Size.String()
+		imageStatus.Unpacked = vdRef.Spec.PersistentVolumeClaim.Size.String()
+		vi.Status.Size = imageStatus
+
+		vi.Status.Format = imageformat.FormatRAW
+		vi.Status.Progress = "100%"
+		vi.Status.Target.PersistentVolumeClaim = dv.Status.ClaimName
+	default:
+		log.Info("Provisioning to PVC is in progress", "dvProgress", dv.Status.Progress, "dvPhase", dv.Status.Phase, "pvcPhase", pvc.Status.Phase)
+
+		vi.Status.Progress = ds.diskService.GetProgress(dv, vi.Status.Progress, service.NewScaleOption(0, 100))
+		vi.Status.Target.PersistentVolumeClaim = dv.Status.ClaimName
+
+		err = ds.diskService.Protect(ctx, vi, dv, pvc)
+		if err != nil {
+			return false, err
+		}
+
+		err = setPhaseConditionForPVCProvisioningImage(ctx, dv, vi, pvc, condition, ds.diskService)
+		if err != nil {
+			return false, err
+		}
+
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (ds ObjectRefVirtualDisk) CleanUpSupplements(ctx context.Context, vi *virtv2.VirtualImage) (bool, error) {
+	supgen := supplements.NewGenerator(common.VIShortName, vi.Name, vi.Namespace, vi.UID)
+
+	importerRequeue, err := ds.importerService.CleanUpSupplements(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+
+	diskRequeue, err := ds.diskService.CleanUpSupplements(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+
+	return importerRequeue || diskRequeue, nil
+}
+
+func (ds ObjectRefVirtualDisk) CleanUp(ctx context.Context, vi *virtv2.VirtualImage) (bool, error) {
+	supgen := supplements.NewGenerator(common.VIShortName, vi.Name, vi.Namespace, vi.UID)
+
+	importerRequeue, err := ds.importerService.CleanUp(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+
+	diskRequeue, err := ds.diskService.CleanUp(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+
+	return importerRequeue || diskRequeue, nil
+}
+
+func (ds ObjectRefVirtualDisk) getEnvSettings(vi *virtv2.VirtualImage, sup *supplements.Generator) *importer.Settings {
+	var settings importer.Settings
+	importer.ApplyBlockDeviceSourceSettings(&settings)
+	importer.ApplyDVCRDestinationSettings(
+		&settings,
+		ds.dvcrSettings,
+		sup,
+		ds.dvcrSettings.RegistryImageForVI(vi),
+	)
+
+	return &settings
+}
+
+func (ds ObjectRefVirtualDisk) Validate(ctx context.Context, vi *virtv2.VirtualImage) error {
+	if vi.Spec.DataSource.ObjectRef == nil || vi.Spec.DataSource.ObjectRef.Kind != virtv2.VirtualImageObjectRefKindVirtualDisk {
+		return fmt.Errorf("not a %s data source", virtv2.ClusterVirtualImageObjectRefKindVirtualDisk)
+	}
+
+	vd, err := ds.diskService.GetVirtualDisk(ctx, vi.Spec.DataSource.ObjectRef.Name, vi.Namespace)
+	if err != nil {
+		return err
+	}
+
+	if vd == nil || vd.Status.Phase != virtv2.DiskReady {
+		return NewVirtualDiskNotReadyError(vi.Spec.DataSource.ObjectRef.Name)
+	}
+
+	if len(vd.Status.AttachedToVirtualMachines) != 0 {
```

Comment:
I like it better when` > 0`
```suggestion
	if len(vd.Status.AttachedToVirtualMachines) > 0 {
```

Reply:
trololo